Refine guidance on helper methods

AArnott opened the pull request.

AArnott committed:
Refine guidance on helper methods

AArnott commented:
@vbfox: I'm interested in your review of this.

vbfox commented:
That sound right, the other cases where it can be useful are :

* First/next -> enumerable helpers handling the specific error return for "no more data" cleanly.
* Maybe adding to the first point that the common cases are growing a buffer until it's big enough and the double call with the first one to get the size and the second one to get the buffer.

I'm still a little bit uncomfortable about the fact that there is 2 type of methods in the same "Name" like `Kernel32`, some that don't throw on error and some that throw without any easy way from the API consumer to do the distinction. But maybe we can do that via documentation.

AArnott committed:
Add reasons to add helper methods

AArnott commented:
Good points. Thanks. I've added them. 
Yes, I'm not quite satisfied with it myself. In fact I just removed [a helper method](https://github.com/AArnott/pinvoke/commit/31bdb443efd2c3be235388b43811e59e77a9ff3b#diff-86bffeaa5625f51ba87e346645c7e07aL177) that was dangerously similar to its p/invoke overload except in the return value.
